### ISSUE TITLE
Windows support

### ISSUE THREAD

**Issue by filerun** [box/brainy#18]
You recommend dropping the constant DIRECTORY_SEPARATOR and always use a forward slash. The code will be easier to read and the library will work on any system, including Windows. It currently fails on Windows because it tried to remove compiled template files without escaping the backslash, thus not forming correct paths.

P.S. The quick start guide is outdated. It took me a while to find out "new Box\Brainy\Brainy()".

**Comment by mattbasta** [comment]
Thanks for filing this issue. Can you explain what you mean by this? 

> It currently fails on Windows because it tried to remove compiled template files without escaping the backslash, thus not forming correct paths.

Can you provide an error message, stack trace, or steps to reproduce the problem? I'm happy to look into what's happening. Thanks!

**Comment by filerun** [comment]
Step 1: `composer require box\brainy`
Step 2. Create test.html with contents: `Hello {$hello|escape}!`
Step 3. Run this:

```
require('vendor/autoload.php');
$b = new Box\Brainy\Brainy();
$b->setCompileDir('compiled');
$b->assign('hello', 'world');
$b->display('test.html');
```

Although it works, it throws this: Warning: unlink(compiled\5017803b9ee9b00cc52db4a18a64b71cfc076fd7.file.test.html.php): No such file or directory in H:\www\temp\brainy\vendor\box\brainy\src\Brainy\Templates\Template.php on line 330

I am not sure I understand why it would try to remove the just compiled file.

**Comment by mattbasta** [comment]
I can give you an answer to that :)

https://github.com/box/brainy/blob/master/src/Brainy/Templates/Template.php#L320

Basically, Brainy writes all new files to a temp file with a random name, then moves the file where it should be. This is done to avoid issues with different file systems and OSs, where the behavior of two processes writing to the same file can differ wildly.

On Windows, the code tries to `unlink()` whatever is at the path that the temp file is being moved to. This prevents the `rename()` operation from failing. This is only done on Windows because PHP behaves differently on Windows than it does on *nix systems. Nothing is failing, it's just throwing a warning (which is not ideal).

@vshen94: Do you want to wrap that call with a `file_exists()` check? Or I can try to do it sometime this weekend.

**Comment by filerun** [comment]
Why not write the data directly to the final location and avoid the filesystem operations?

**Comment by mattbasta** [comment]
Another process might have compiled the file, but there's no guarantee that it's compiled the same version of the file. The compilation process takes a non-zero amount of time, so you can run into a dangerous race condition if the template changes.

**Comment by filerun** [comment]
So what happens when there are 100 subsequent requests during the time the first request is compiling the template? Each request starts its own compilation? I would have thought that they would check if a compilation already started (detected by some kind of locking mechanism) and wait for it.

**Comment by mattbasta** [comment]
Unfortunately PHP doesn't give you a way to do that without installing a C extension. There's also the case where compilation fails, or goes on virtually forever. That could take a server down, or even a whole Datacenter under the right conditions.

**Comment by filerun** [comment]
Compiling a template file would take normally just some microseconds. Consider this scenario:
- HTTP request #1 triggers the template compilation, creates the target file, with a name something like "compiled.php". Next to the file, a temporary file named "compiled_lock.php"
- HTTP request #2 requests the use of the same template, before starting to parse & compile anything, looks for the file "compiled_lock.php". If there, means the compilation is under progress. Checks the "filemtime()" of the lock file and if the lock file is not older than a second, "usleep" for 50 microseconds, clearstatcache() and verify again if the lock file is present. If the lock is present, sleep again, and repeat process a few times until one of the following applies:
- If the lock is no longer present it means the compilation progress started by request #1 has been completed, thus load the compiled file.
- If the lock file gets older than a second (or some other preconfigured time out), assumes compilation failed so the template is broken, do not attempt to do anything with it, throw an error or whatever.

All the subsequent requests will do the same. No costly file manipulation, no race condition, pretty good performance in the case of a busy website with frequent template changing. Everybody happy :)

**Comment by filerun** [comment]
Imagine 10.000 requests on a page with a large template that has a problem somewhere towards its bottom. 10.000 PHP processes parsing the large template file up to the problem, instead of just avoid any futile work and usage of server resources. The above solution covers this problem as well.

**Comment by mattbasta** [comment]
The compilation can take far longer than microseconds. Large applications commonly take hundreds of milliseconds or more, which is why it's advisable to precompile your templates.

**Comment by filerun** [comment]
Even better for those kind of cases. Let's imagine it takes 2 seconds to compile it. What happens in the background for a page like that which has hundreds of visitors at the same time. Instead for all those subsequent requests to hang for less than 2 seconds for the first request to finish the job for everybody, just serve them garbage or errors?

I use the above technique in a very similar situation, when the requested PHP script needs to generate (compile) a thumbnail for a very large file, and there are many other requests made at the same time for the same thumbnail. It works great.

**Comment by mattbasta** [comment]
There's no way of safely knowing whether another request is already compiling the template. You might, for instance, be on a network filesystem with high latency. Consequently, you can't rely on lock files. 

Any approach that polls the filesystem is a nonstarter because a very busy server will quickly be overtaken with disk IO. Since PHP has no built in way of getting notifications for file changes, there is no solution that will work at scale. Since this happens at server startup, it's particularly egregious because you've presumably just released new code, and problems with the new code might be masked by problems with compilation (and vise versa). 

If every request compiles its own templates, you eliminate the possibility of race conditions or a slow process blocking the whole server. Yes, it does lead to wasted work for a short time, but it prevents a whole class of very real issues. 

Imagine you have a request which starts compiling a template and the process OOMs. If another 100 requests are waiting, they'll only ever error out. There's no (reasonable) way of electing a new request to start the compilation. By using the approach we have, requests that wouldn't otherwise OOM will still be able to succeed. 

The warning that you're seeing will be fixed by putting a guard around it. This is happening because Smarty mutes warnings, and Box doesn't run Windows servers. Nobody has tested that code path in practice since the project was forked. We'll make sure it gets fixed for you. :)

**Comment by filerun** [comment]
Well, you're putting the issue under the rug. It's no longer about Windows.

Your argument regarding disk IO doesn't stand, as the alternative creates much worse disk usage, without any benefit whatsoever. Compare the performance of 1000 "file_exists" calls to 1000 full template parsing.
And for the case where the compilation fails, I already gave you the solution. The new request that attempts the compilation again will be the first one that finds out that the process timed out (by checking the creation time of the lock file).
This also improves the user experience, as it is always better to have a page that takes longer (by miliseconds) to load than have one that shows a Brainy error ;)

Just picture things like this: you have 10.000 people trying to go through one door. Should you allow all of them to get in front and press on the handle, if the first one that gets there already tried and can tell the rest that the door it's locked (and it's trying to unlock it)? :)

Brainy seems like a great thing, that's why I am taking the time to try to help, but the first moment I gave it a try a pretty nasty error popped up, which showed me how inefficiently things work behind the scene. This can be done smarter than in Smarty ;)
